src/raftstore/coprocessor/region_info_accessor.rs

```diff
@@ -0,0 +1,724 @@
+            assert_eq!(removed_region.get_id(), region.get_id());
+
+            // Ignore if it's a stale message
+            let removed_version = removed_region.get_region_epoch().get_version();
+            let removed_conf = removed_region.get_region_epoch().get_conf_ver();
+            if removed_version > region.get_region_epoch().get_version()
+                || removed_conf > region.get_region_epoch().get_conf_ver()
+            {
+                // It shouldn't be deleted. Put it back.
+                removed_region_info.region = removed_region;
+                self.regions.insert(region.get_id(), removed_region_info);
+                return;
+            }
+
+            assert_eq!(removed_region.get_id(), region.get_id());
+            let end_key = data_end_key(removed_region.get_end_key());
+
+            let removed_id = self.region_ranges.remove(&end_key).unwrap();
+            assert_eq!(removed_id, region.get_id());
+        } else {
+            // It's possible that the region is already removed because it's end_key is used by
+            // another newer region.
+            debug!(
+                "region_collection: destroying region {} but it doesn't exist",
+                region.get_id()
+            )
+        }
+    }
+
+    fn handle_role_change(&mut self, region: Region, new_role: StateRole) {
+        let region_id = region.get_id();
+        if self.regions.get(&region_id).is_none() {
+            warn!("region_collection: role change on region {} but the region doesn't exist. create it.", region_id);
+            self.create_region(region);
+        }
+
+        if let Some(r) = self.regions.get_mut(&region_id) {
+            r.role = new_role;
+        }
+    }
+
+    fn handle_raftstore_event(&mut self, event: RaftStoreEvent) {
+        match event {
+            RaftStoreEvent::CreateRegion { region } => {
+                self.handle_create_or_update_region(region);
+            }
+            RaftStoreEvent::UpdateRegion { region } => {
+                self.handle_create_or_update_region(region);
+            }
+            RaftStoreEvent::DestroyRegion { region } => {
+                self.handle_destroy_region(region);
+            }
+            RaftStoreEvent::RoleChange { region, role } => {
+                self.handle_role_change(region, role);
+            }
+        }
+    }
+}
+
+impl Runnable<RegionCollectionMsg> for RegionCollection {
+    fn run(&mut self, task: RegionCollectionMsg) {
+        match task {
+            RegionCollectionMsg::RaftStoreEvent(event) => {
+                self.handle_raftstore_event(event);
+            }
+            RegionCollectionMsg::DebugDump(tx) => {
+                tx.send((self.regions.clone(), self.region_ranges.clone()))
+                    .unwrap();
+            }
+        }
+    }
+}
+
+/// `RegionInfoAccessor` keeps all region information separately from raftstore itself.
+#[derive(Clone)]
+pub struct RegionInfoAccessor {
```

Comment:
How about RegionInfoProvider?

Reply:
The name `RegionInfoProvider` is already used by a trait

Reply:
I think naming it `Accessor` is ok. It's like in some languages we say a method `get_xxx` is an accessor, and `get_xxx` is a modifier